src/services/fileStorage/strategies/awsS3.js

```diff
@@ -1,46 +1,105 @@
 const filePermissionHelper = require('../utils/filePermissionHelper');
 const { removeLeadingSlash } = require('../utils/filePathHelper');
 
-// const prodMode = process.env.NODE_ENV === 'production';
+const AbstractFileStorageStrategy = require('./interface.js');
+
+const getCorsRules = () => ([{
+	AllowedHeaders: ['*'],
+	AllowedMethods: ['PUT'],
+	AllowedOrigins: [process.env.HOST],
+	MaxAgeSeconds: 300,
+}]);
+
+const getConfig = (provider) => {
+	const awsConfig = new aws.Config({
+		signatureVersion: 'v4',
+		s3ForcePathStyle: true,
+		sslEnabled: true,
+		accessKeyId: provider.accessKeyId,
+		secretAccessKey: provider.secretAccessKey,
+		region: provider.region || 'eu-de',
+		endpointUrl: provider.endpointUrl,
+		cors_rules: getCorsRules(),
+	});
+	awsConfig.endpoint = new aws.Endpoint(provider.endpointUrl);
+	return awsConfig;
+};
+
+const chooseProvider = async (schoolId) => {
+	const provider = await storageProviderModel.aggregate([
+		{ $match: { isShared: true } },
+		{ $addFields: { freeBuckets: { $subtract: ['$maxBuckets', { $size: '$schools' }] } } },
+		{ $sort: { freeBuckets: -1 } },
+		{ $limit: 1 },
+	]).exec();
+	if (!Array.isArray(provider) || provider.length === 0) throw new Error('No available provider found.');
+	await storageProviderModel.findByIdAndUpdate(provider[0]._id,
+		{ $push: { schools: schoolId } },
+		{ safe: true, upsert: true }).lean().exec();
+	return provider;
+};
 
+const FEATURE_MULTIPLE_S3_PROVIDERS_ENABLED = Configuration.get('FEATURE_MULTIPLE_S3_PROVIDERS_ENABLED');
+// begin legacy
 let awsConfig = {};
-try {
-	//	awsConfig = require(`../../../../config/secrets.${prodMode ? 'js' : 'json'}`).aws;
-	/* eslint-disable global-require, no-unused-expressions */
-	(['production'].includes(process.env.NODE_ENV))
-		? awsConfig = require('../../../../config/secrets.js').aws
-		: awsConfig = require('../../../../config/secrets.json').aws;
-	/* eslint-enable global-require, no-unused-expressions */
-} catch (e) {
-	logger.warning('The AWS config couldn\'t be read');
+if (!FEATURE_MULTIPLE_S3_PROVIDERS_ENABLED) {
+	try {
+		//	awsConfig = require(`../../../../config/secrets.${prodMode ? 'js' : 'json'}`).aws;
+		/* eslint-disable global-require, no-unused-expressions */
+		(['production'].includes(process.env.NODE_ENV))
+			? awsConfig = require('../../../../config/secrets.js').aws
+			: awsConfig = require('../../../../config/secrets.json').aws;
+		/* eslint-enable global-require, no-unused-expressions */
+	} catch (e) {
+		logger.warning('The AWS config couldn\'t be read');
+	}
 }
+// end legacy
+
+const createAWSObject = async (schoolId) => {
+	const school = await schoolModel.findOne({ _id: schoolId }).lean().exec();
+	if (school === null) throw new NotFound('School not found.');
+
+	if (FEATURE_MULTIPLE_S3_PROVIDERS_ENABLED) {
+		const S3_KEY = Configuration.get('S3_KEY');
+		let provider = await storageProviderModel.findOne({ schools: schoolId }).lean().exec();
```

Comment:
how to take care, a school is mentioned only once in a provider? should the property storage provider not be added on schools?